File: Blind 75/07 - Trees/102-binary-tree-level-order-traversal.py
'''
Given the root of a binary tree, return the level order traversal of its nodes' values. (i.e., from left to right, level by level).
'''
import collections
from time import time
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        
        res, q = [], [root]
        while q:
            layer_values, next_layer = [], []
            
            for node in q:
                layer_values.append(node.val)
                if node.left:
                    next_layer.append(node.left)
                if node.right:
                    next_layer.append(node.right)
            
            q = next_layer
            res.append(layer_values)
        
        return res

    def reference(self, root: Optional[TreeNode]) -> List[List[int]]:
        res = []
        q = collections.deque()
        if root:
            q.append(root)

        while q:
            val = []

            for i in range(len(q)):
                node = q.popleft()
                val.append(node.val)
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            res.append(val)
        return res

    def quantify(self, test_cases, runs=100000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.levelOrder(case))
                else:
                    self.levelOrder(case)
        print(f'Runtime for our solution: {time() - sol_start}')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case))
                else:
                    self.reference(case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [
        TreeNode(3,
                 TreeNode(9),
                 TreeNode(20, TreeNode(15), TreeNode(7)),
                 ),
        TreeNode(1),
        None,
        # Additional
        TreeNode(1, TreeNode(2)),
        TreeNode(1,
                 TreeNode(2, TreeNode(4)),
                 TreeNode(3, None, TreeNode(5))
                 )
    ]
    test.quantify(test_cases)

File: Blind 75/07 - Trees/JavaScript/102-binary-tree-level-order-traversal.js
const { performance } = require('perf_hooks');

class TreeNode {
  constructor(val, left, right) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

class Solution {
  method(root) {
    if (!root) {
      return [];
    }

    let res = [];
    let queue = [[root, 0]];

    while (queue.length > 0) {
      const [node, level] = queue.shift();
      if (res[level] !== undefined) {
        res[level].push(node.val);
      } else {
        res[level] = [node.val];
      }

      if (node.left) {
        queue.push([node.left, level + 1]);
      }
      if (node.right) {
        queue.push([node.right, level + 1]);
      }
    }

    return res;
  }

  reference(root, level = 0, levels = []) {
    const isBaseCase = root === null;
    if (isBaseCase) return levels;

    const isLastNode = level === levels.length;
    if (isLastNode) levels.push([]);

    levels[level].push(root.val);

    return this.dfs(root, level, levels);
  }

  dfs = (root, level, levels) => {
    if (root.left) this.reference(root.left, level + 1, levels);
    if (root.right) this.reference(root.right, level + 1, levels);

    return levels;
  };

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(input));
        else this.method(input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(input));
        else this.reference(input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  new TreeNode(
    3,
    new TreeNode(9),
    new TreeNode(20, new TreeNode(15), new TreeNode(7))
  ),
  new TreeNode(1),
  null,
  // Additional
  new TreeNode(1, new TreeNode(2)),
  new TreeNode(
    1,
    new TreeNode(2, new TreeNode(4)),
    new TreeNode(3, null, new TreeNode(5))
  ),
];
test.quantify(testCases);
